network/src/validator_network/mod.rs

```diff
@@ -139,3 +148,181 @@ impl<TMessage: Message + Default> FusedStream for NetworkEvents<TMessage> {
         self.inner.is_terminated()
     }
 }
+
+/// `NetworkSender` is the generic interface from upper network applications to
```

Comment:
One common pattern for implementing such APIs is to keep these functions as the primary API of the underlying entity itself, in this case the `NetworkProvider` instead of creating a new struct. The `channel::Sender<NetworkRequest>` can be stored privately in the struct. We probably can't make that change just yet, since we'd need to migrate all the `NetworkProvider` clients to also use this API, but what are your thoughts on doing this eventually?

Reply:
The whole point was to extract these APIs into an ownable struct so you can make `&mut` references to the channel and send without cloning all the time. We wouldn't be able to do this if everyone held an `Arc<NetworkProvider>` instead. Neither would we be able to monomorphize the functions so they specialize to each protobuf message type, which would require passing around `Box<dyn prost::Message>` instead. Maybe I'm misunderstanding what you're proposing?

Reply:
yeah, I think both are fair points.